sklearn/metrics/tests/test_pairwise.py

```diff
@@ -556,6 +556,38 @@ def test_euclidean_distances():
                                   Y_norm_squared=np.zeros_like(Y_norm_sq))
     assert_greater(np.max(np.abs(wrong_D - D1)), .01)
 
+    # Check euclidean_distances when using float32 for one or both arguments
+    X32 = X.astype(np.float32)
+    Y32 = Y.astype(np.float32)
+    D64 = euclidean_distances(X, Y)
+    D64_32 = euclidean_distances(X, Y32)
+    D32_64 = euclidean_distances(X32, Y)
+    D32_32 = euclidean_distances(X32, Y32)
+    assert_equal(D64_32.dtype, np.float64)
+    assert_equal(D32_64.dtype, np.float64)
+    assert_equal(D32_32.dtype, np.float32)
+    assert_array_almost_equal(D64_32, D64)
+    assert_array_almost_equal(D32_64, D64)
+    assert_array_almost_equal(D32_32, D64)
+
+    # Check that {X,Y}_norm_squared are used with float32 arguments
+    X32_norm_sq = 0.5 * (X32 ** 2).sum(axis=1).reshape(1, -1)
+    Y32_norm_sq = 0.5 * (Y32 ** 2).sum(axis=1).reshape(1, -1)
+    DYN = euclidean_distances(X32, Y32, Y_norm_squared=Y32_norm_sq)
+    DXN = euclidean_distances(X32, Y32, X_norm_squared=X32_norm_sq)
+    DXYN = euclidean_distances(X32, Y32, X_norm_squared=X32_norm_sq,
+                               Y_norm_squared=Y32_norm_sq)
+    assert_greater(np.max(np.abs(DYN - D64)), .01)
+    assert_greater(np.max(np.abs(DXN - D64)), .01)
+    assert_greater(np.max(np.abs(DXYN - D64)), .01)
+
+    # Check that the accuracy with float32 is not too bad
+    X = np.array([[0.9215765222065525, 0.9682577158572608],
+                  [0.9221151778782808, 0.9681831844652774]])
+    D64 = euclidean_distances(X)
+    D32 = euclidean_distances(X.astype(np.float32))
+    assert_array_almost_equal(D32, D64)
```

Comment:
On master I get,
```py
In [5]: D64 = euclidean_distances(X)

In [6]: D64
Out[6]: 
array([[0.        , 0.00054379],
       [0.00054379, 0.        ]])

In [7]: D32 = euclidean_distances(X.astype(np.float32))

In [8]: D32
Out[8]: 
array([[0.        , 0.00059802],
       [0.00059802, 0.        ]], dtype=float32)
```

and `assert_array_almost_equal` uses 6 decimals by default, so since this passes, it would increase the accuracy from 1 to 3 significant digits. Maybe use `assert_allclose` with `rtol` to make it more explicit?